core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+
+            // If there is no longer a common prefix, there is only one key in the TraverseInfo,
+            // and then all remaining siblings are required for the multiproof.
+            if curr_traverse_info.path_slice_lower == curr_traverse_info.path_slice_upper {
+                let path_proof = &path_proofs[curr_traverse_info.path_slice_lower];
+                let sub_path_proof = SubPathProof {
+                    terminal: path_proof.terminal.clone(),
+                    // The depth at which the terminal starts not sharing any siblings
+                    depth: curr_traverse_info.path_index as u8,
+                    inner_siblings: path_proof
+                        .siblings
+                        .iter()
+                        .skip(curr_traverse_info.path_index)
+                        .copied()
+                        .collect(),
+                };
+                sub_paths.push(sub_path_proof);
+
+                // insert all collected siblings in this sub tree in ext_sibling
+                // at the position this bisection is pointing at
+                for sibling in ext_sibling_in_sub_tree.into_iter().rev() {
+                    external_siblings.insert(curr_traverse_info.ext_sibling_index, sibling);
+                }
+                ext_sibling_in_sub_tree = vec![];
+
+                // skip to the next bisection in the stak, if empty we're finished
+                curr_traverse_info = match stack_traverse_info.pop() {
+                    Some(i) => i,
+                    None => break,
+                };
+                continue;
+            }
+
+            // check if at least two key_path in the key_paths slice
+            // has two different bits in position path_index
+            //
+            // they are ordered and all the bits up to path_index
+            // are shared so we can just check the first and the last one differs
+            let path_lower = path_proofs[curr_traverse_info.path_slice_lower]
+                .terminal
+                .key_path()
+                .view_bits::<Msb0>();
+            let path_upper = path_proofs[curr_traverse_info.path_slice_upper]
+                .terminal
+                .key_path()
+                .view_bits::<Msb0>();
+
+            if path_lower[curr_traverse_info.path_index]
+                != path_upper[curr_traverse_info.path_index]
+            {
+                // if they differ we can skip their siblings but we need to bisec the slice
+                //
+                // binary search between key_paths in the slice to see where to
+                // perform the bisection
+
+                // path_slice_upper and path_slice_lower will never be the same otherwise
+                // bits at path_index would have been the same
+
+                let mut low = curr_traverse_info.path_slice_lower;
+                let mut up = curr_traverse_info.path_slice_upper;
+                let mut mid;
+                loop {
```

Comment:
Yes, you're totally right. That's an example of premature optimization, where I thought that since the vector was sorted, I could have done it better than the compiler.

Using `slice::binary_search` or one of its variations is not doable here because the first 1 is required.

As you said, a solution like this seems to be the cleanest and most reliable one (and it could be improved later on)

```rust
                let mid = curr_traverse_info.lower
                    + path_proofs[curr_traverse_info.lower..=curr_traverse_info.upper]
                        .iter()
                        .position(|path_proof| {
                            path_proof.terminal.key_path().view_bits::<Msb0>()
                                [curr_traverse_info.path_bit_index]
                        })
                        .expect("There must be at least one 1 bit");

```